### DISCUSSION TITLE
The use of rem inside the Fozzie components style

### DISCUSSION THREAD

**Discussion by semenkivje** [justeattakeaway/fozzie-components#1264] [General]
For discussion: Fozzie components could be used inside projects which are not using Fozzie styles package.
In that case, we couldn't rely on rem value for sizes or fonts because this value could be another on the project level.

**Comment by ashleynolan** [comment]
So just to clarify – is the issue that the base font size is set to `16px`, or that the resulting rem values are calculated against that base font size?

I'm guessing the latter, which would make it awkward to change, as the way we bundle components through webpack means that it's the resulting CSS that gets imported by consuming applications rather than SCSS, which is easier to change.

If you can confirm though, as it's hard to know without seeing the application (unless you can ping me a link to look at the site too – best to do that via Slack).

**Reply by semenkivje** [reply]
@ashleynolan I created for you a simple example - https://codesandbox.io/s/blissful-joliot-emydq?file=/src/App.vue 
So as you can see in button styles - the calculated font size is 18.2858 px.

I think it happened because of the @font-size mixing from @justeat/f-utils here:
https://github.com/justeat/fozzie-components/blob/master/packages/components/atoms/f-button/src/components/Button.vue#L282

So now we have the same situation in our project with the f-button: it doesn't look like in vue.pie.design (https://vue.pie.design/?path=/story/components-atoms-f-button--button-component)

**Reply by ashleynolan** [reply]
Ok, that makes sense.

So I think this is basically highlighting an assumption we've made with the components, that consuming apps will set their font-size to be 14px, as that's what we do in the fozzie base framework.

The main issue we have is that in the components we can't make this value dynamic currently – the styles have to be compiled out to CSS, so they are completely static at that point.

I think there are potentially two solutions to this issue:

1. Generate two sets of static CSS that could be imported by consumers. One which uses the default browser sizes (which will set these rem values as if the base is the browser default – 16px) and one which will use the fozzie base font-size for users of that framework. So then, people using the component can either use the default browser sizing or the fozzie specific sizing. This will be easy to do in fozzie itself – we would just need to look into if it's possible to build out 2 versions of the CSS in each component.
2. Move fozzie and the components over to use CSS variables, so that we can use a dynamic variable for the base font-size value in fozzie. This would mean that consuming applications would then be able to change the base font size to whatever they wanted and it would ripple through the components. 

My preferred approach would definitely be `2`. We're dropping IE11 support imminently, and so switching over to use CSS variables for this should be possible at this point. It won't be something we can do super quickly though – we're likely talking at least the start of October until we can get these tickets into our backlog due to the Icing updates being made and released.

TL:DR – I think the issue should be solvable, but it's not something we can just update immediately as it's a reasonably big shift in how the framework is currently set-up.

We could potentially look at if option 1 would work int he interim, but it just depends on how quickly you need the change. I think we'd be able to start looking at updating the framework to do Option 2 in a few weeks depending on other priorities.

**Reply by DamianMullins** [reply]
Related to this, I've noticed that a lot of the PIE designs in Figma use a base font-size of 16px — was there any particular reason we're using 14px in Fozzie?

**Reply by ashleynolan** [reply]
`14px` was our default paragraph `font-size` I think previously –  I think that's why it was lift and shifted when we moved to Icing.

It does look like it's changed now to be `16px`, but we've just kept the 14px default, so we may actually just be able to include this as part of our `v6` fozzie updates potentially which would be quicker.

@dandel10n Just looping you in to this too. Does this look right to you too? If so, we should likely create a ticket to update this to 16px (or the body paragraph design-token variable that's set to 16px) as that will also solve the issue partner centre are having.

**Reply by dandel10n** [reply]
Yes, I checked and can confirm that base font size is 16px now. I missed this change when was working on font updates... I will create a ticket for this change as part of v6 fozzie updates :+1:

**Reply by semenkivje** [reply]
Generating separate versions is probably not the most flexible idea. Since there may be projects with really custom value for rem. As I know some projects can use 62.5% as described here https://www.sitepoint.com/understanding-and-using-rem-units-in-css/
In this case, we will have to generate versions of the styles for each of them.

Are there any opportunities to remove rem usage at the component level? 
Some UI libraries do not use them at the component level, for example
Vuetify - https://vuetifyjs.com/en/components/buttons/#usage
ElementUI https://element.eleme.io/#/en-US/component/button

What do you think about this?

**Reply by ashleynolan** [reply]
@semenkivje I think using rem's should likely be fine when we move more permanently to CSS variables (which will happen in the next couple of months I think). As I understand it (needs a little more of a spike to make sure) this should then mean that any consuming app can set their base font size to whatever they want and it'll dynamically be reflected throughout the framework.

I think the main next steps will solve the short-term issue PC is seeing – we'll move fozzie over to use 16px base font size, and then the components will be updated to reflect this too. Then in the long-term, we'll move the framework over to use CSS variables, which should mean that we have full flexibility over setting the base font size inside any consuming application.

@dandel10n has already made a ticket to update the base font size, so that should get done next week. We'll let you know once this is also reflected inside components.  As it's a breaking change, it'll be a part of v6 of fozzie (which includes the new Icing updates for type and colours), unless we also back-port this into v5 as well (we can do that, but you'll eventually need to upgrade to get the latest versions of components, as we won't be able to keep supporting this version once we fully release v6.

So there may be some consideration over this rollout with @chrisfirth86, to ensure that when you do integrate with this new version, the styling will be in-line with the new `JETSans` font and Icing colours.

**Reply by semenkivje** [reply]
Thanks for the information and help from your team @ashleynolan.

About the fonts.
Perhaps this is worthy of a separate discussion - but for me, it's related as we use some global staff on atoms level: 
I have checked and for example, the button component has font definitions at the level of its styles.
But wouldn't it be easier to update styles(site font) - if we rely on the site font without declaring them inside atoms components? 

For example, if we will have some new components in the next version of Fozzy and they will have another font family. In this case, we should update the version of all atoms components and our site styles to use one of them.

**Reply by ashleynolan** [reply]
@semenkivje Yeh – me and @dandel10n were actually talking about this last week too. I think the font declarations you mention need to be removed too.

They are there from when we used to have different heading and body font families (back when we used to use – now that we use `JetSans` for both headings and body copy, we can delete these references and just inherit the base font as you mention. We'll be fixing this shortly (or feel free to put a PR in to remove it if you need it before we have chance to update it).

**Reply by dandel10n** [reply]
@semenkivje hi, I removed font declaration from the button component and the base font has been updated to 16px as well :+1:
